eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/PlaceSuggestionModule.java

```diff
@@ -0,0 +1,202 @@
+
+
+/**
+ * {@link SearchModule} for showing the Google Place API places suggestions.
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class PlaceSuggestionModule implements SearchModule, PlaceSuggestionItem.OnClickListener
+{
+    public static SelectableSearchModuleFactory<GeoPlace> FACTORY = new SelectableSearchModuleFactory<GeoPlace>()
+    {
+        @Override
+        public SearchModule create(Activity activity, NonNotifyingChangeableList changeableList, ListItemSelectionAction<GeoPlace> itemSelectionAction)
+        {
+            GoogleApiClient googleApiClient = new GoogleApiClient
+                    .Builder(activity)
+                    .addApi(Places.GEO_DATA_API)
+                    .addApi(Places.PLACE_DETECTION_API)
+                    .build();
+            /**
+             * TODO enableAutomanage OR manual error handling
+             * enableAutoManage() on the builder would enable automatic default error handling as well,
+             * but it's tricky to get initialization correctly with Activity and retained Fragment lifecycles.
+             * Either enable automanage or add 'manual' error handling with addOnConnectionFailedListener().
+             *
+             * See https://developers.google.com/android/reference/com/google/android/gms/common/api/GoogleApiClient.Builder.html#enableAutoManage(android.support.v4.app.FragmentActivity, com.google.android.gms.common.api.GoogleApiClient.OnConnectionFailedListener)
+             */
+
+            googleApiClient.connect();
+            return new PlaceSuggestionModule(changeableList, itemSelectionAction, googleApiClient);
+        }
+
+    };
+
+    private static final String TAG = PlaceSuggestionModule.class.getSimpleName();
+
+    private final NonNotifyingChangeableList<ListItem> mChangeableList;
+
+    private ListItemSelectionAction<GeoPlace> mSelectionAction;
+
+    private GoogleApiClient mGoogleApiClient;
+    private ExecutorService mExecutorService;
+
+    private PlaceSuggestionQueryTask.Client mSuggestionQueryTaskClient;
+    private String mLastQuery;
+
+    private PlaceByIdTaskClient mPlaceByIdTaskClient;
+
+
+    private PlaceSuggestionModule(NonNotifyingChangeableList<ListItem> changeableList,
+                                  ListItemSelectionAction<GeoPlace> itemSelectionAction,
+                                  GoogleApiClient googleApiClient)
+    {
+        mChangeableList = changeableList;
+        mSelectionAction = itemSelectionAction;
+        mGoogleApiClient = googleApiClient;
+
+        mExecutorService = Executors.newSingleThreadExecutor();
+        mSuggestionQueryTaskClient = new PlaceSuggestionQueryTaskClient();
+        mPlaceByIdTaskClient = new PlaceByIdTaskClient();
+    }
+
+
+    @Override
+    public void shutDown()
+    {
+        mGoogleApiClient.disconnect();
+
+        // TODO think about this, is callback better? That would have to be nulled out as well.
+        mSelectionAction = null;
```

Comment:
Ok, I don't bother with nulling out the action (without using the modules in a retained fragment it cannot actually cause context leak.)

Thanks for the note on `WeakReference` usage, updated the `AsycnTasks`.